pkg/object/mqttproxy/session.go

```diff
@@ -0,0 +1,245 @@
+	s.Lock()
+	defer s.Unlock()
+	s.info.EGName = egName
+	s.info.Name = name
+	s.store()
+}
+
+func (s *Session) subscribe(topics []string, qoss []byte) error {
+	s.Lock()
+	defer s.Unlock()
+	for i, t := range topics {
+		s.info.Topics[t] = int(qoss[i])
+	}
+	s.store()
+	return nil
+}
+
+func (s *Session) unsubscribe(topics []string) error {
+	s.Lock()
+	defer s.Unlock()
+	for _, t := range topics {
+		delete(s.info.Topics, t)
+	}
+	s.store()
+	return nil
+}
+
+func (s *Session) allSubscribes() ([]string, []byte, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	var sub []string
+	var qos []byte
+	for k, v := range s.info.Topics {
+		sub = append(sub, k)
+		qos = append(qos, byte(v))
+	}
+	return sub, qos, nil
+}
+
+func (s *Session) getPacketFromMsg(topic string, payload []byte, qos byte) *packets.PublishPacket {
+	p := packets.NewControlPacket(packets.Publish).(*packets.PublishPacket)
+	p.Qos = qos
+	p.TopicName = topic
+	p.Payload = payload
+	p.MessageID = s.nextID
+	// the overflow is okay here
+	// the session will give unique id from 0 to 65535 and do this again and again
+	s.nextID++
+	return p
+}
+
+func (s *Session) publish(topic string, payload []byte, qos byte) {
+	client := s.broker.getClient(s.info.ClientID)
+	s.Lock()
+	defer s.Unlock()
+
+	if client == nil {
+		logger.Errorf("client %s is offline", s.info.ClientID)
+	} else {
+		p := s.getPacketFromMsg(topic, payload, qos)
+		if qos == Qos0 {
+			go client.writePacket(p)
+		} else if qos == Qos1 {
+			msg := getMsg(topic, payload, qos)
+			s.pending[p.MessageID] = msg
+			s.pendingQueue = append(s.pendingQueue, p.MessageID)
+			go client.writePacket(p)
+		} else {
+			logger.Errorf("mqtt.publish: current not support to publish message with qos=2")
+		}
+	}
+}
+
+func (s *Session) puback(p *packets.PubackPacket) {
+	s.Lock()
+	defer s.Unlock()
+	delete(s.pending, p.MessageID)
+}
+
+func (s *Session) cleanSession() bool {
+	return s.info.CleanFlag
+}
+
+func (s *Session) close() {
+	close(s.done)
+}
+
+func (s *Session) doResend() {
+	client := s.broker.getClient(s.info.ClientID)
+	s.Lock()
+	defer s.Unlock()
+
+	if len(s.pending) == 0 {
+		s.pendingQueue = []uint16{}
+		return
+	}
+	for i, idx := range s.pendingQueue {
+		if val, ok := s.pending[idx]; ok {
+			// find first msg need to resend
+			s.pendingQueue = s.pendingQueue[i:]
+			p := packets.NewControlPacket(packets.Publish).(*packets.PublishPacket)
+			p.Qos = byte(val.Qos)
+			p.TopicName = val.Topic
+			payload, err := base64.StdEncoding.DecodeString(val.B64Payload)
+			if err != nil {
+				logger.Errorf("mqtt.doResend: base64 decode error for Message B64Payload %s", err)
+				return
+			}
+			p.Payload = payload
+			p.MessageID = idx
+			if client != nil {
+				go client.writePacket(p)
+			}
+			return
+		}
+	}
+}
+
+func (s *Session) backgroundResendPending() {
+	for {
+		select {
+		case <-s.done:
+			return
+		default:
+			s.doResend()
+		}
+		time.Sleep(100 * time.Millisecond)
```

Comment:
```suggestion
                ticker := time.NewTicker(100 * time.Millisecond)
		select {
		case <-s.done:
			return
		case <-ticker.C:
			s.doResend()
		}
```

this could make the function return quickly